app/models/manageiq/providers/amazon/agent_coordinator.rb

```diff
@@ -0,0 +1,423 @@
+    return false unless bucket.exists?
+
+    obj_id = "#{heartbeat_prefix}#{agent_id}"
+    obj = bucket.object(obj_id)
+    return false unless obj.exists?
+
+    last_heartbeat = obj.last_modified
+    _log.debug("#{obj.key}: Last heartbeat time stamp: #{last_heartbeat}")
+
+    Time.now.utc - last_heartbeat < interval && ec2.instance(agent_id).state.name == "running"
+  rescue => err
+    _log.error("#{agent_id}: #{err.message}")
+    false
+  end
+
+  def activate_agents
+    agent_ids.each do |id|
+      agent = ec2.instance(id)
+      if agent.state.name == "stopped"
+        agent.start
+        agent.wait_until_running
+        _log.info("Agent #{id} is activated to serve requests.")
+        return id
+      else
+        _log.warn("Agent #{id} is in abnormal state: #{agent.state.name}.")
+        next
+      end
+    end
+
+    _log.error("Failed to activate agents: #{agent_ids}.")
+  end
+
+  def deploy_agent
+    _log.info("Deploying agent ...")
+    @deploying = true
+
+    kp = find_or_create_keypair
+    zone_name = ec2.client.describe_availability_zones.availability_zones[0].zone_name
+    subnets = get_subnets(zone_name)
+    raise "No subnet_id is available for #{zone_name}!" if subnets.empty?
+    security_group_id = find_or_create_security_group(subnets[0].vpc_id)
+    find_or_create_profile
+
+    instance                = ec2.create_instances(
+      :image_id             => get_agent_image_id,
+      :min_count            => 1,
+      :max_count            => 1,
+      :key_name             => kp.name,
+      :security_group_ids   => [security_group_id],
+      :instance_type        => 't2.micro',
+      :placement            => {
+        :availability_zone => zone_name
+      },
+      :subnet_id            => subnets[0].subnet_id,
+      :iam_instance_profile => {
+        :name => label
+      },
+      :tag_specifications   => [{
+        :resource_type => "instance",
+        :tags          => [{
+          :key   => "Name",
+          :value => label
+        }]
+      }]
+    )
+    ec2.client.wait_until(:instance_status_ok, :instance_ids => [instance[0].id])
+
+    _log.info("Start to load SSA application, this may take a while ...")
+
+    setup_agent(instance[0])
+    _log.info("SSA agent is ready to receive requests.")
+
+    instance[0].id
+  end
+
+  def setup_agent(instance)
+    # register if needed
+    register(instance)
+
+    # prepare work directory
+    ssh_commands(instance, ["sudo mkdir -p #{WORK_DIR}; sudo chmod go+w #{WORK_DIR}"], agent_ami_login_user)
+    # scp the default setting yaml file
+    create_config_yaml("config.yml")
+    scp_file("config.yml", WORK_DIR.to_s, instance, agent_ami_login_user)
+
+    # run docker image
+    command_line = "sudo docker run -d --restart=always -v /dev:/host_dev -v #{WORK_DIR}/config.yml:#{WORK_DIR}/config.yml --privileged #{agent_docker_name}"
+    ssh_commands(instance, [command_line.to_s], agent_ami_login_user)
+  end
+
+  # TODO: for downstream
```

Comment:
Remove this.  If anything, just remove the method entirely and change the caller to `register(instance) if respond_to?(:register)`

Reply:
Actually, a better way to handle this is to actually implement it in upstream by registering when the credentials are present.  That way, on upstream, no credentials will be present, so it will be a no-op and downstream the credentials will be required.  In fact you could make the "are credentials required" a configurable value...something like `:agent_ami_registration_required: true`

Reply:
To build on this... For any private docker registries, you need to `docker login`.  So I would call the method something like
```ruby
def login_to_registry
  return unless settings.docker_registry.url
  command_line = "docker login"
  command_line << " #{settings.docker_registry.url}"
  command_line << " -u #{settings.docker_registry.username} -p #{settings.docker_registry.password}" if settings.docker_registry.username
  send_command_over_ssh(command_line)
end

Reply:
@bdunne For docker login, what's the username and password for our manageiq?

Reply:
There is no username/password for public docker images.